### PR TITLE
create activity selected image detail

### PR DESCRIPTION
**Pull request by anhpham2403**
https://dev.framgia.com/redmine/issues/152456

### CODE REVIEW

File: app/src/main/java/com/framgia/englishconversation/screen/imageDetail/OnMediaModelItemTouchListener.java [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package com.framgia.englishconversation.screen.imageDetail;
```
Comment (DoanVanToan) [inline app/src/main/java/com/framgia/englishconversation/screen/imageDetail/OnMediaModelItemTouchListener.java]: ERROR: Name 'com.framgia.englishconversation.screen.imageDetail' must match pattern '^[a-z]+(\.[a-z][a-z0-9_]{1,})*$'.

File: app/src/main/java/com/framgia/englishconversation/screen/imageDetail/OnMediaModelItemTouchListener.java [hunk 2]
```diff
@@ -0,0 +1,9 @@
...
+
+/**
+ * Created by VinhTL on 21/12/2017.
+ */
+
+public interface OnMediaModelItemTouchListener<T> {
```
Comment (DoanVanToan) [inline app/src/main/java/com/framgia/englishconversation/screen/imageDetail/OnMediaModelItemTouchListener.java]: WARNING: Type Javadoc comment is missing an @param <T> tag.

File: app/src/main/java/com/framgia/englishconversation/widget/ZoomableImageView.java [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+import android.graphics.Bitmap;
+import android.graphics.Matrix;
+import android.graphics.PointF;
+import android.support.v7.widget.AppCompatImageView;
+import android.util.AttributeSet;
+import android.view.MotionEvent;
+import android.view.ScaleGestureDetector;
+import android.view.View;
+
+public class ZoomableImageView extends AppCompatImageView {
```
Comment (DoanVanToan) [inline app/src/main/java/com/framgia/englishconversation/widget/ZoomableImageView.java]: WARNING: Missing a Javadoc comment.

File: app/src/main/java/com/framgia/englishconversation/widget/ZoomableImageView.java [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+    float width, height;
+    float saveScale = 1f;
+    float right, bottom, origWidth, origHeight, bmWidth, bmHeight;
+
+    ScaleGestureDetector scaleDetector;
+    Context context;
+
+
+
+    public ZoomableImageView(Context context, AttributeSet attr) {
```
Comment (DoanVanToan) [inline app/src/main/java/com/framgia/englishconversation/widget/ZoomableImageView.java]: WARNING: Method length is 79 lines (max allowed is 60).

File: app/src/main/java/com/framgia/englishconversation/widget/ZoomableImageView.java [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+        this.context = context;
+        scaleDetector = new ScaleGestureDetector(context, new ScaleListener());
+        matrix.setTranslate(1f, 1f);
+        m = new float[9];
+        setImageMatrix(matrix);
+        setScaleType(ScaleType.MATRIX);
+
+        setOnTouchListener(new OnTouchListener() {
+
+            @Override
```
Comment (DoanVanToan) [inline app/src/main/java/com/framgia/englishconversation/widget/ZoomableImageView.java]: WARNING: Method length is 64 lines (max allowed is 60).

File: app/src/main/java/com/framgia/englishconversation/widget/ZoomableImageView.java [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+        super.setImageBitmap(bm);
+        bmWidth = bm.getWidth();
+        bmHeight = bm.getHeight();
+    }
+
+    public void setMaxZoom(float x) {
+        maxScale = x;
+    }
+
+    private class ScaleListener extends ScaleGestureDetector.SimpleOnScaleGestureListener {
```
Comment (DoanVanToan) [inline app/src/main/java/com/framgia/englishconversation/widget/ZoomableImageView.java]: WARNING: Missing a Javadoc comment.

File: app/src/main/java/com/framgia/englishconversation/widget/ZoomableImageView.java [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+                matrix.postScale(mScaleFactor, mScaleFactor, width / 2, height / 2);
+                if (mScaleFactor < 1) {
+                    matrix.getValues(m);
+                    float x = m[Matrix.MTRANS_X];
+                    float y = m[Matrix.MTRANS_Y];
+                    if (mScaleFactor < 1) {
+                        if (Math.round(origWidth * saveScale) < width) {
+                            if (y < -bottom) {
+                                matrix.postTranslate(0, -(y + bottom));
+                            } else if (y > 0) matrix.postTranslate(0, -y);
```
Comment (DoanVanToan) [inline app/src/main/java/com/framgia/englishconversation/widget/ZoomableImageView.java]: WARNING: 'if' construct must use '{}'s.

File: app/src/main/java/com/framgia/englishconversation/widget/ZoomableImageView.java [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+                        } else {
+                            if (x < -right) {
+                                matrix.postTranslate(-(x + right), 0);
+                            } else if (x > 0) matrix.postTranslate(-x, 0);
```
Comment (DoanVanToan) [inline app/src/main/java/com/framgia/englishconversation/widget/ZoomableImageView.java]: WARNING: 'if' construct must use '{}'s.

File: app/src/main/java/com/framgia/englishconversation/widget/ZoomableImageView.java [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+            } else {
+                matrix.postScale(mScaleFactor, mScaleFactor, detector.getFocusX(),
+                        detector.getFocusY());
+                matrix.getValues(m);
+                float x = m[Matrix.MTRANS_X];
+                float y = m[Matrix.MTRANS_Y];
+                if (mScaleFactor < 1) {
+                    if (x < -right) {
+                        matrix.postTranslate(-(x + right), 0);
+                    } else if (x > 0) matrix.postTranslate(-x, 0);
```
Comment (DoanVanToan) [inline app/src/main/java/com/framgia/englishconversation/widget/ZoomableImageView.java]: WARNING: 'if' construct must use '{}'s.

File: app/src/main/java/com/framgia/englishconversation/widget/ZoomableImageView.java [hunk 10]
```diff
@@ -0,0 +1,213 @@
...
+                    if (y < -bottom) {
+                        matrix.postTranslate(0, -(y + bottom));
+                    } else if (y > 0) matrix.postTranslate(0, -y);
```
Comment (DoanVanToan) [inline app/src/main/java/com/framgia/englishconversation/widget/ZoomableImageView.java]: WARNING: 'if' construct must use '{}'s.

File: app/src/main/java/com/framgia/englishconversation/screen/imageDetail/OnMediaModelItemTouchListener.java (hunk 1 above)
Comment (DoanVanToan) [inline app/src/main/java/com/framgia/englishconversation/screen/imageDetail/OnMediaModelItemTouchListener.java]: WARNING: [PackageCase] Package name contains upper case characters
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/naming.html#PackageCase

File: app/src/main/java/com/framgia/englishconversation/widget/ZoomableImageView.java (hunk 7 above)
Comment (DoanVanToan) [inline app/src/main/java/com/framgia/englishconversation/widget/ZoomableImageView.java]: WARNING: [IfStmtsMustUseBraces] Avoid using if statements without curly braces
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/braces.html#IfStmtsMustUseBraces

File: app/src/main/java/com/framgia/englishconversation/widget/ZoomableImageView.java (hunk 8 above)
Comment (DoanVanToan) [inline app/src/main/java/com/framgia/englishconversation/widget/ZoomableImageView.java]: WARNING: [IfStmtsMustUseBraces] Avoid using if statements without curly braces
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/braces.html#IfStmtsMustUseBraces

File: app/src/main/java/com/framgia/englishconversation/widget/ZoomableImageView.java (hunk 9 above)
Comment (DoanVanToan) [inline app/src/main/java/com/framgia/englishconversation/widget/ZoomableImageView.java]: WARNING: [IfStmtsMustUseBraces] Avoid using if statements without curly braces
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/braces.html#IfStmtsMustUseBraces

File: app/src/main/java/com/framgia/englishconversation/widget/ZoomableImageView.java (hunk 10 above)
Comment (DoanVanToan) [inline app/src/main/java/com/framgia/englishconversation/widget/ZoomableImageView.java]: WARNING: [IfStmtsMustUseBraces] Avoid using if statements without curly braces
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/braces.html#IfStmtsMustUseBraces

File: app/src/main/res/drawable/ic_left_arrow.xml [hunk 11]
```diff
@@ -0,0 +1,4 @@
+<vector android:height="24dp" android:viewportHeight="129.0"
```
Comment (DoanVanToan) [inline app/src/main/res/drawable/ic_left_arrow.xml:1]: WARNING: [UnusedResources] The resource `R.drawable.ic_left_arrow` appears to be unused
 Unused resources make applications larger and slow down builds.

File: app/src/main/res/drawable/ic_right_arrow.xml [hunk 12]
```diff
@@ -0,0 +1,4 @@
+<vector android:height="24dp" android:viewportHeight="129.0"
```
Comment (DoanVanToan) [inline app/src/main/res/drawable/ic_right_arrow.xml:1]: WARNING: [UnusedResources] The resource `R.drawable.ic_right_arrow` appears to be unused
 Unused resources make applications larger and slow down builds.

File: app/src/main/java/com/framgia/englishconversation/widget/ZoomableImageView.java [hunk 13]
```diff
@@ -0,0 +1,213 @@
...
+        scaleDetector = new ScaleGestureDetector(context, new ScaleListener());
+        matrix.setTranslate(1f, 1f);
+        m = new float[9];
+        setImageMatrix(matrix);
+        setScaleType(ScaleType.MATRIX);
+
+        setOnTouchListener(new OnTouchListener() {
+
+            @Override
+            public boolean onTouch(View v, MotionEvent event) {
```
Comment (DoanVanToan) [inline app/src/main/java/com/framgia/englishconversation/widget/ZoomableImageView.java]: WARNING: [ClickableViewAccessibility] `com/framgia/englishconversation/widget/ZoomableImageView$1#onTouch` should call `View#performClick` when a click is detected
 If a `View` that overrides `onTouchEvent` or uses an `OnTouchListener` does not also implement `performClick` and call it when clicks are detected, the `View` may not handle accessibility actions properly. Logic handling the click actions should ideally be placed in `View#performClick` as some accessibility services invoke `performClick` when a click action should occur.

File: app/src/main/java/com/framgia/englishconversation/widget/ZoomableImageView.java [hunk 14]
```diff
@@ -0,0 +1,235 @@
...
+
+    float redundantXSpace, redundantYSpace;
+    float width, height;
+    float saveScale = 1f;
+    float right, bottom, origWidth, origHeight, bmWidth, bmHeight;
+
+    ScaleGestureDetector scaleDetector;
+    Context context;
+
+    public ZoomableImageView(Context context, AttributeSet attr) {
```
Comment (DoanVanToan) [inline app/src/main/java/com/framgia/englishconversation/widget/ZoomableImageView.java]: WARNING: Method length is 81 lines (max allowed is 60).

File: app/src/main/java/com/framgia/englishconversation/widget/ZoomableImageView.java [hunk 15]
```diff
@@ -0,0 +1,235 @@
...
+        this.context = context;
+        scaleDetector = new ScaleGestureDetector(context, new ScaleListener());
+        matrix.setTranslate(1f, 1f);
+        m = new float[9];
+        setImageMatrix(matrix);
+        setScaleType(ScaleType.MATRIX);
+
+        setOnTouchListener(new OnTouchListener() {
+
+            @Override
```
Comment (DoanVanToan) [inline app/src/main/java/com/framgia/englishconversation/widget/ZoomableImageView.java]: WARNING: Method length is 66 lines (max allowed is 60).

File: app/src/main/res/drawable/ic_right_arrow.xml (hunk 12 above)
Comment (anhpham2403) [reply app/src/main/res/drawable/ic_right_arrow.xml:1]: em dùng cho pull sau

File: app/src/main/res/drawable/ic_left_arrow.xml (hunk 11 above)
Comment (anhpham2403) [reply app/src/main/res/drawable/ic_left_arrow.xml:1]: em dùng cho pull sau

File: app/src/main/java/com/framgia/englishconversation/widget/ZoomableImageView.java [hunk 16]
```diff
@@ -0,0 +1,235 @@
...
+import android.view.MotionEvent;
+import android.view.ScaleGestureDetector;
+import android.view.View;
+
+/**
+ * ZoomableImageView
+ */
+
+public class ZoomableImageView extends AppCompatImageView {
+    Matrix matrix = new Matrix();
```
Comment (DoanVanToan) [inline app/src/main/java/com/framgia/englishconversation/widget/ZoomableImageView.java]: e self review lại class này nhé, thiếu access modifier ở tất cả các variable

File: app/src/main/java/com/framgia/englishconversation/utils/binding/BindingUtils.java [hunk 17]
```diff
@@ -99,6 +99,15 @@ public static void loadImage(final ImageView imageView, Uri uri) {
                 .into(imageView);
     }
 
+    @BindingAdapter("bind:imageUrlFullSize")
```
Comment (fs-sournary) [inline app/src/main/java/com/framgia/englishconversation/utils/binding/BindingUtils.java:102]: databinding cái này có rồi mà anh

File: app/src/main/java/com/framgia/englishconversation/widget/ZoomableImageView.java [hunk 18]
```diff
@@ -0,0 +1,212 @@
...
+    private Matrix mMatrix = new Matrix();
+    private PointF mLast = new PointF();
+    private PointF mStart = new PointF();
+    private float[] mValues;
+    private float mRedundantXSpace, mRedundantYSpace;
+    private float mWidth, mHeight, mOrigWidth, mOrigHeight, mBmWidth, mBmHeight, mRight, mBottom;
+    private float mSaveScale = 1f;
+    private ScaleGestureDetector mScaleDetector;
+
+    public ZoomableImageView(Context context, AttributeSet attr) {
```
Comment (DoanVanToan) [inline app/src/main/java/com/framgia/englishconversation/widget/ZoomableImageView.java:33]: WARNING: Method length is 79 lines (max allowed is 60).

File: app/src/main/java/com/framgia/englishconversation/widget/ZoomableImageView.java [hunk 19]
```diff
@@ -0,0 +1,212 @@
...
+        super.setClickable(false);
+        mScaleDetector = new ScaleGestureDetector(context, new ScaleListener());
+        mMatrix.setTranslate(1f, 1f);
+        mValues = new float[9];
+        setImageMatrix(mMatrix);
+        setScaleType(ScaleType.MATRIX);
+
+        setOnTouchListener(new OnTouchListener() {
+
+            @Override
```
Comment (DoanVanToan) [inline app/src/main/java/com/framgia/englishconversation/widget/ZoomableImageView.java]: WARNING: Method length is 65 lines (max allowed is 60).

File: app/src/main/java/com/framgia/englishconversation/utils/binding/BindingUtils.java (hunk 17 above)
Comment (anhpham2403) [reply app/src/main/java/com/framgia/englishconversation/utils/binding/BindingUtils.java:102]: cái kia có centerCrop() còn cái này ko có để có thể zoom image

File: app/src/main/java/com/framgia/englishconversation/screen/imagedetail/OnMediaModelItemTouchListener.java [hunk 20]
```diff
@@ -0,0 +1,9 @@
+package com.framgia.englishconversation.screen.imagedetail;
+
+/**
+ * Created by VinhTL on 21/12/2017.
+ */
+
+public interface OnMediaModelItemTouchListener<T> {
```
Comment (DoanVanToan) [inline app/src/main/java/com/framgia/englishconversation/screen/imagedetail/OnMediaModelItemTouchListener.java]: WARNING: Type Javadoc comment is missing an @param <T> tag.

File: app/src/main/java/com/framgia/englishconversation/widget/ZoomableImageView.java [hunk 21]
```diff
@@ -0,0 +1,213 @@
...
+    private Matrix mMatrix = new Matrix();
+    private PointF mLast = new PointF();
+    private PointF mStart = new PointF();
+    private float[] mValues;
+    private float mRedundantXSpace, mRedundantYSpace;
+    private float mWidth, mHeight, mOrigWidth, mOrigHeight, mBmWidth, mBmHeight, mRight, mBottom;
+    private float mSaveScale = 1f;
+    private ScaleGestureDetector mScaleDetector;
+
+    public ZoomableImageView(Context context,
```
Comment (DoanVanToan) [inline app/src/main/java/com/framgia/englishconversation/widget/ZoomableImageView.java]: WARNING: Method length is 79 lines (max allowed is 60).

File: app/src/main/java/com/framgia/englishconversation/widget/ZoomableImageView.java [hunk 22]
```diff
@@ -0,0 +1,213 @@
...
+        super(context, attr);
+        super.setClickable(false);
+        mScaleDetector = new ScaleGestureDetector(context, new ScaleListener());
+        mMatrix.setTranslate(1f, 1f);
+        mValues = new float[9];
+        setImageMatrix(mMatrix);
+        setScaleType(ScaleType.MATRIX);
+        setOnTouchListener(new OnTouchListener() {
+
+            @Override
```
Comment (DoanVanToan) [inline app/src/main/java/com/framgia/englishconversation/widget/ZoomableImageView.java]: WARNING: Method length is 65 lines (max allowed is 60).

File: app/src/main/java/com/framgia/englishconversation/widget/ZoomableImageView.java [hunk 23]
```diff
@@ -0,0 +1,213 @@
...
+    private Matrix mMatrix = new Matrix();
+    private PointF mLast = new PointF();
+    private PointF mStart = new PointF();
+    private float[] mValues;
+    private float mRedundantXSpace, mRedundantYSpace;
+    private float mWidth, mHeight, mOrigWidth, mOrigHeight, mBmWidth, mBmHeight, mRight, mBottom;
+    private float mSaveScale = 1f;
+    private ScaleGestureDetector mScaleDetector;
+
+    public ZoomableImageView(Context context, AttributeSet attr) {
```
Comment (DoanVanToan) [inline app/src/main/java/com/framgia/englishconversation/widget/ZoomableImageView.java:33]: WARNING: Method length is 62 lines (max allowed is 60).

File: app/src/main/java/com/framgia/englishconversation/widget/ZoomableImageView.java [hunk 24]
```diff
@@ -0,0 +1,208 @@
...
+                mSaveScale = MIN_SCALE;
+                mScaleFactor = MIN_SCALE / origScale;
+            }
+            mRight = mWidth * mSaveScale - mWidth - (2 * mRedundantXSpace * mSaveScale);
+            mBottom = mHeight * mSaveScale - mHeight - (2 * mRedundantYSpace * mSaveScale);
+            mMatrix.getValues(mValues);
+            if (mOrigWidth * mSaveScale <= mWidth || mOrigHeight * mSaveScale <= mHeight) {
+                mMatrix.postScale(mScaleFactor, mScaleFactor, mWidth / 2, mHeight / 2);
+                if (mScaleFactor < 1) {
+                    if (mScaleFactor < 1) {
```
Comment (DoanVanToan) [inline app/src/main/java/com/framgia/englishconversation/widget/ZoomableImageView.java]: WARNING: [CollapsibleIfStatements] These nested if statements could be combined
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/basic.html#CollapsibleIfStatements

File: app/src/main/java/com/framgia/englishconversation/widget/ZoomableImageView.java [hunk 25]
```diff
@@ -0,0 +1,207 @@
...
+            }
+            mRight = mWidth * mSaveScale - mWidth - (2 * mRedundantXSpace * mSaveScale);
+            mBottom = mHeight * mSaveScale - mHeight - (2 * mRedundantYSpace * mSaveScale);
+            if (mOrigWidth * mSaveScale <= mWidth || mOrigHeight * mSaveScale <= mHeight) {
+                mMatrix.postScale(mScaleFactor, mScaleFactor, mWidth / 2, mHeight / 2);
+                mMatrix.getValues(mValues);
+                float x = mValues[Matrix.MTRANS_X];
+                float y = mValues[Matrix.MTRANS_Y];
+                if (mScaleFactor < 1) {
+                    if (mScaleFactor < 1) {
```
Comment (DoanVanToan) [inline app/src/main/java/com/framgia/englishconversation/widget/ZoomableImageView.java]: WARNING: [CollapsibleIfStatements] These nested if statements could be combined
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/basic.html#CollapsibleIfStatements

**Review by TheVinhLuong — APPROVED** [review]

**Pull request merged**